src/build.c

```diff
@@ -879,40 +879,38 @@ static GPid build_spawn_cmd(GeanyDocument *doc, const gchar *cmd, const gchar *d
 }
 
 
-/* Returns: NULL if there was an error, or the working directory the script was created in.
- * vte_cmd_nonscript is the location of a string which is filled with the command to be used
- * when vc->skip_run_script is set, otherwise it will be set to NULL */
-static gchar *prepare_run_script(GeanyDocument *doc, gchar **vte_cmd_nonscript, guint cmdindex)
+/* Returns: NULL if there was an error, or the command to be executed. If Geany is
+ * set to use a run script, the returned value is a path to the script that runs
+ * the command; otherwise the command itself is returned. working_dir is a pointer
+ * to the working directory from which the command is executed. Both strings are
+ * in the locale encoding. */
+static gchar *prepare_run_cmd(GeanyDocument *doc, gchar **working_dir, guint cmdindex)
 {
 	GeanyBuildCommand *cmd = NULL;
-	gchar *working_dir = NULL;
 	const gchar *cmd_working_dir;
 	gboolean autoclose = FALSE;
-	gboolean result = FALSE;
-	gchar *tmp;
-	gchar *cmd_string;
+	gchar *cmd_string_utf8, *working_dir_utf8, *run_cmd;
 	GError *error = NULL;
 
-	if (vte_cmd_nonscript != NULL)
-		*vte_cmd_nonscript = NULL;
-
 	cmd = get_build_cmd(doc, GEANY_GBG_EXEC, cmdindex, NULL);
 
-	cmd_string = build_replace_placeholder(doc, cmd->command);
+	cmd_string_utf8 = build_replace_placeholder(doc, cmd->command);
 	cmd_working_dir =  cmd->working_dir;
 	if (EMPTY(cmd_working_dir))
 		cmd_working_dir = "%d";
-	working_dir = build_replace_placeholder(doc, cmd_working_dir); /* in utf-8 */
+	working_dir_utf8 = build_replace_placeholder(doc, cmd_working_dir);
+	if (!EMPTY(working_dir_utf8))
```

Comment:
It might return NULL if doc == NULL or it isn't valid. Still, you are right the fi doesn't have to be there since utils_get_locale_from_utf8() works for NULL strings too.